pkg/deviceplugin/topology.go

```diff
@@ -0,0 +1,254 @@
+		if err != nil {
+			return nil, err
+		}
+
+		for _, hint := range hints {
+			for _, nNode := range strings.Split(hint.NUMAs, ",") {
+				nNodeID, err := strconv.ParseInt(strings.TrimSpace(nNode), 10, 64)
+				if err != nil {
+					return nil, errors.Wrapf(err, "unable to convert numa node %s into int64", nNode)
+				}
+				if !nodeIDs[nNodeID] {
+					result.Nodes = append(result.Nodes, &pluginapi.NUMANode{ID: nNodeID})
+					nodeIDs[nNodeID] = true
+				}
+			}
+		}
+	}
+	sort.Slice(result.Nodes, func(i, j int) bool { return result.Nodes[i].ID < result.Nodes[j].ID })
+	return &result, nil
+}
+
+// NewTopologyHints return array of hints for the device and its slaves (e.g. RAID).
+func NewTopologyHints(devPath string) (hints TopologyHints, err error) {
+	hints = make(TopologyHints)
+	realDevPath, err := filepath.EvalSymlinks(devPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed get realpath for %s", devPath)
+	}
+	for p := realDevPath; strings.HasPrefix(p, mockRoot+"/sys/devices/"); p = filepath.Dir(p) {
+		hint := TopologyHint{Provider: p}
+		fileMap := map[string]*string{
+			"local_cpulist": &hint.CPUs,
+			"numa_node":     &hint.NUMAs,
+		}
+		if err = readFilesInDirectory(fileMap, p); err != nil {
+			return nil, err
+		}
+		// Workarounds for broken information provided by kernel
+		if hint.NUMAs == "-1" {
+			// non-NUMA aware device or system, ignore it
+			hint.NUMAs = ""
+		}
+		if hint.NUMAs != "" && hint.CPUs == "" {
+			// broken topology hint. BIOS reports socket id as NUMA node
+			// First, try to get hints from parent device or bus.
+			parentHints, er := NewTopologyHints(filepath.Dir(p))
+			if er == nil {
+				cpulist := map[string]bool{}
+				numalist := map[string]bool{}
+				for _, h := range parentHints {
+					if h.CPUs != "" {
+						cpulist[h.CPUs] = true
+					}
+					if h.NUMAs != "" {
+						numalist[h.NUMAs] = true
+					}
+				}
+				if cpus := strings.Join(mapKeys(cpulist), ","); cpus != "" {
+					hint.CPUs = cpus
+				}
+				if numas := strings.Join(mapKeys(numalist), ","); numas != "" {
+					hint.NUMAs = numas
+				}
+			}
+			// if after parent hints we still don't have CPUs hints, use numa hint as sockets.
+			if hint.CPUs == "" && hint.NUMAs != "" {
+				hint.Sockets = hint.NUMAs
+				hint.NUMAs = ""
+			}
+		}
+		if hint.CPUs != "" || hint.NUMAs != "" || hint.Sockets != "" {
+			hints[hint.Provider] = hint
+			break
+		}
+	}
+	slaves, _ := filepath.Glob(filepath.Join(realDevPath, "slaves/*"))
+	for _, slave := range slaves {
+		slaveHints, er := NewTopologyHints(slave)
+		if er != nil {
+			return nil, er
+		}
+		hints = MergeTopologyHints(hints, slaveHints)
+	}
+	return
+}
+
+// MergeTopologyHints combines org and hints.
+func MergeTopologyHints(org, hints TopologyHints) (res TopologyHints) {
+	if org != nil {
+		res = org
+	} else {
+		res = make(TopologyHints)
+	}
+	for k, v := range hints {
+		if _, ok := res[k]; ok {
+			continue
+		}
+		res[k] = v
+	}
+	return
```

Comment:
in modern Go recommendations for short functions it is actually opposite. If the function fits into one screen (~10-15 lines), it is better to use naked returns to improve readability (less polluted).

Reply:
I checked https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters 
> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

I wouldn't interpret it as a recommendation for naked returns.